### PR TITLE
Power port

### PR DESCRIPTION
**Pull request by brihh**
changes for the Power port.

### CODE REVIEW

File: as/src/fabric/clustering.c [hunk 1]
```diff
@@ -1348,7 +1348,8 @@ if (((size) > STACK_ALLOC_LIMIT()) && buffer) {cf_free(buffer);}
 #define TRACE_LOG(context, format, ...)
 #endif
 
-#define CF_TRACE CF_FAULT_SEVERITY_UNDEF
+//bri//#define CF_TRACE CF_FAULT_SEVERITY_UNDEF
```
Comment (grooverdan) [inline as/src/fabric/clustering.c]: leave comments of prior code out of the commit.

File: .gitmodules [hunk 2]
```diff
@@ -1,7 +1,3 @@
-[submodule "modules/common"]
-	path = modules/common
-	url = https://github.com/aerospike/aerospike-common.git
-	ignore = dirty
```
Comment (grooverdan) [inline .gitmodules:4]: not appropriate to upstream

File: .gitmodules [hunk 3]
```diff
@@ -21,3 +21,6 @@
 	path = modules/jemalloc
 	url = https://github.com/aerospike/jemalloc.git
 	ignore = dirty
+[submodule "modules/common"]
```
Comment (grooverdan) [inline .gitmodules:24]: not appropriate to upstream

File: as/src/fabric/clustering.c (hunk 1 above)
Comment (grooverdan) [reply as/src/fabric/clustering.c]: commit message should be more explicit, and its not a 'hack', its a propper fix'

File: make_in/Makefile.in [hunk 4]
```diff
@@ -75,6 +75,11 @@ ifneq ($(ARCH),$(filter $(ARCH),ppc64 ppc64le))
   COMMON_CFLAGS += -march=nocona
 endif
 
+# The type char is unsigned by default on PPC -- we want signed
+ifeq ($(ARCH),$(filter $(ARCH),ppc64 ppc64le))
+  COMMON_CFLAGS += -fsigned-char
```
Comment (grooverdan) [inline make_in/Makefile.in:80]: fix by making the code where the `char` needs to be `signed`, and make that as a `signed char`

File: cf/src/hardware.c [hunk 5]
```diff
@@ -376,6 +376,7 @@ read_index(const char *path, uint16_t *val)
 	char *end;
 	uint64_t x = strtoul(buff, &end, 10);
 
+	x = (x / 8) & 0x0F; // hack for Power numbering
```
Comment (grooverdan) [inline cf/src/hardware.c:414]: not portable. Fix correctly

File: as/src/fabric/clustering.c [hunk 6]
```diff
@@ -1336,21 +1336,6 @@ if (((size) > STACK_ALLOC_LIMIT()) && buffer) {cf_free(buffer);}
 #define DEBUG(format, ...) cf_debug(AS_CLUSTERING, format, ##__VA_ARGS__)
 #define DETAIL(format, ...) cf_detail(AS_CLUSTERING, format, ##__VA_ARGS__)
 
-#ifdef TRACE_ENABLED
```
Comment (grooverdan) [inline as/src/fabric/clustering.c]: bad commit message. unclear why you've removed this. Couldn't you just build without `TRACE_ENABLED` being defined.

File: as/src/base/index.c [hunk 7]
```diff
@@ -676,15 +692,14 @@ as_index_sprig_get_insert_vlock(as_index_sprig *isprig, uint8_t tree_id,
 
 		cf_arenax_handle t_h = isprig->sprig->root_h;
 		as_index *t = RESOLVE_H(t_h);
+		_mm_prefetch(t, _MM_HINT_NTA);
```
Comment (grooverdan) [inline as/src/base/index.c:711]: commit message should also explain that the prefetch is happening as early as possible.

File: as/src/base/index.c [hunk 8]
```diff
@@ -31,7 +31,23 @@
...
 #include <xmmintrin.h>
+#elif defined(__powerpc64__)
+// cache prefetch transient
+static inline void
+_mm_prefetch(void *pointer, int unused __attribute__((unused)))
+{
+  __asm__ ( " dcbtt     0,%0;" : : "b" (pointer));
+}
+#else
+#define _mm_prefetch(X, Y) __builtin_prefetch(X)
```
Comment (grooverdan) [inline as/src/base/index.c:45]: _MM_HINT_NTA implies that it is a non-temporarl fetch, therefore `0` can be used as the second argument to __builtin_prefetch. Could check the ISA2.07 to see about dcbtt and see if it also is a non-temporal fetch? Maybe the __builtin_prefetch is sufficient on POWER too.

File: as/src/base/thr_demarshal.c [hunk 9]
```diff
@@ -436,7 +436,7 @@ thr_demarshal(void *unused)
 
 		// Iterate over all events.
 		for (i = 0; i < nevents; i++) {
-			cf_socket *ssock = events[i].data;
+			cf_socket *ssock = events[i].data.ptr;
```
Comment (grooverdan) [inline as/src/base/thr_demarshal.c:439]: commit message could say we're using the standard provided union of auxiliary data as per the epoll_ctl man page.

**Review by grooverdan — CHANGES_REQUESTED** [review]
these commits should be slit into separate pull requests and the non-upstreamable commits removed.

File: as/src/Makefile [hunk 10]
```diff
@@ -95,7 +95,9 @@ INCLUDES += -I$(COMMON)/target/$(PLATFORM)/include
 INCLUDES += -I$(MOD_LUA)/target/$(PLATFORM)/include
 INCLUDES += -I$(JANSSON)/src
 INCLUDES += -I$(S2)
-INCLUDES += -I$(XDR_INCLUDES)
+ifneq ($(XDR_INCLUDES), )
```
Comment (grooverdan) [inline as/src/Makefile:98]: Something like `ifdef` would be a bit more direct:
<pre>
ifdef S2
  INCLUDES += -I$(S2)
endif
ifdef XDR_INCLUDES
  INCLUDES += -I$(XDR_INCLUDES)
endif
</pre>

File: as/src/base/index.c (hunk 8 above)
Comment (brihh) [reply as/src/base/index.c:45]: will you make the recommendation on this? this change was part of your original commit:

https://github.com/grooverdan/aerospike-server/commit/e07e3744b94d9da862c663fc479030d48874a213

File: as/src/base/thr_demarshal.c (hunk 9 above)
Comment (brihh) [reply as/src/base/thr_demarshal.c:439]: done.

File: as/src/Makefile (hunk 10 above)
Comment (brihh) [reply as/src/Makefile:98]: done

File: make_in/Makefile.in (hunk 4 above)
Comment (brihh) [reply make_in/Makefile.in:80]: i'm going to leave this as-is for now. if i have more time or if aerospike wants to fix each individual char, then we'll go that route.

File: as/src/base/index.c (hunk 7 above)
Comment (brihh) [reply as/src/base/index.c:711]: done

**Comment by brihh** [comment]
cancelling this pull request - commits updated, i'll do another..bri.

**Pull request closed without merging**